components/operations-controller/controllers/operation_controller.go

```diff
@@ -16,36 +16,420 @@ package controllers
 
 import (
 	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"github.com/kyma-incubator/compass/components/director/pkg/resource"
+	"github.com/kyma-incubator/compass/components/director/pkg/webhook"
+	"github.com/kyma-incubator/compass/components/operations-controller/internal/client"
+	"github.com/kyma-incubator/compass/components/operations-controller/internal/director"
+	"github.com/kyma-incubator/compass/components/operations-controller/internal/tenant"
+	web_hook "github.com/kyma-incubator/compass/components/operations-controller/internal/webhook"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"time"
+
+	"github.com/kyma-incubator/compass/components/director/pkg/graphql"
+
+	corev1 "k8s.io/api/core/v1"
 
 	"github.com/go-logr/logr"
-	"k8s.io/apimachinery/pkg/runtime"
 	ctrl "sigs.k8s.io/controller-runtime"
-	"sigs.k8s.io/controller-runtime/pkg/client"
 
-	operationsv1alpha1 "github.com/kyma-incubator/compass/components/operations-controller/api/v1alpha1"
+	"github.com/kyma-incubator/compass/components/operations-controller/api/v1alpha1"
 )
 
 // OperationReconciler reconciles a Operation object
 type OperationReconciler struct {
-	client.Client
-	Log    logr.Logger
-	Scheme *runtime.Scheme
+	k8sClient      client.Client
+	webhookConfig  *web_hook.Config
+	directorClient director.Client
+	webhookClient  web_hook.Client
+	log            logr.Logger
+}
+
+func NewOperationReconciler(k8sClient client.Client, webhookConfig *web_hook.Config, directorClient director.Client, webhookClient web_hook.Client, logger logr.Logger) *OperationReconciler {
+	return &OperationReconciler{
+		k8sClient:      k8sClient,
+		webhookConfig:  webhookConfig,
+		log:            logger,
+		directorClient: directorClient,
+		webhookClient:  webhookClient,
+	}
 }
 
 // +kubebuilder:rbac:groups=operations.compass,resources=operations,verbs=get;list;watch;create;update;patch;delete
 // +kubebuilder:rbac:groups=operations.compass,resources=operations/status,verbs=get;update;patch
 
 func (r *OperationReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
-	_ = context.Background()
-	_ = r.Log.WithValues("operation", req.NamespacedName)
+	ctx := context.Background()
+	logger := r.log.WithValues("operation", req.NamespacedName)
+
+	operation, err := r.k8sClient.Get(ctx, req.NamespacedName)
+	if err != nil {
+		logger.Error(err, fmt.Sprintf("Unable to retrieve %s from API server", req.NamespacedName))
+		return ctrl.Result{}, nil
+	}
+
+	requestData, err := parseRequestData(operation)
+	if err != nil {
+		logger.Error(err, operationError("Unable to parse request data", operation))
+		return ctrl.Result{}, nil
+	}
+
+	ctx = tenant.SaveToContext(ctx, requestData.TenantID)
```

Comment:
The tenant is dynamic and we have to plug it into the token transport decorator somehow. Using the context seems to be the most streamlined solution.
Otherwise we'd have to propagate the tenantID through numerous functions, even some functions which doesn't make much sense. Also the end goal is to generate a self-signed token with which to communicate with the Director, this sounds like it makes most sense to be near the transport where the header is attached.

https://github.com/kyma-incubator/compass/blob/8d38369c87bee7ad5c693b0e3310b87e0fba54d3/components/operations-controller/internal/director/token_provider.go#L71

Also propagating the tenant throughout the director client functions would look strange in terms of common interfaces as none of the gql client functions from the system broker do it like that.